src/cloudflare/internal/vectorize-api.ts

```diff
@@ -0,0 +1,214 @@
+
+enum Operation {
+  INDEX_GET = 0,
+  VECTOR_QUERY = 1,
+  VECTOR_INSERT = 2,
+  VECTOR_UPSERT = 3,
+  VECTOR_GET = 4,
+  VECTOR_DELETE = 5,
+}
+
+class VectorizeIndexImpl implements VectorizeIndex {
+  public constructor(
+    private readonly fetcher: Fetcher,
+    private readonly indexId: string
+  ) {}
+
+  public async describe(): Promise<VectorizeIndexDetails> {
+    const res = await this._send(
+      Operation.INDEX_GET,
+      `indexes/${this.indexId}`,
+      {
+        method: "GET",
+      }
+    );
+
+    return await toJson<VectorizeIndexDetails>(res);
+  }
+
+  public async query(
+    vector: VectorFloatArray | number[],
+    options: VectorizeQueryOptions
+  ): Promise<VectorizeMatches> {
+    const res = await this._send(
+      Operation.VECTOR_QUERY,
+      `indexes/${this.indexId}/query`,
+      {
+        method: "POST",
+        body: JSON.stringify({
+          ...options,
+          vector: Array.isArray(vector) ? vector : Array.from(vector),
+        }),
+        headers: {
+          "content-type": "application/json",
+          accept: "application/json",
+        },
+      }
+    );
+
+    return await toJson<VectorizeMatches>(res);
+  }
+
+  public async insert(
+    vectors: VectorizeVector[]
+  ): Promise<VectorizeVectorMutation> {
+    const res = await this._send(
+      Operation.VECTOR_INSERT,
+      `indexes/${this.indexId}/insert`,
+      {
+        method: "POST",
+        body: JSON.stringify({
+          vectors: vectors.map((vec) => ({
+            ...vec,
+            values: Array.isArray(vec.values)
+              ? vec.values
+              : Array.from(vec.values),
+          })),
+        }),
+        headers: {
+          "content-type": "application/json",
+          "cf-vector-search-dim-width": String(
+            vectors.length ? vectors[0]?.values?.length : 0
+          ),
+          "cf-vector-search-dim-height": String(vectors.length),
+          accept: "application/json",
+        },
+      }
+    );
+
+    return await toJson<VectorizeVectorMutation>(res);
+  }
+
+  public async upsert(
+    vectors: VectorizeVector[]
+  ): Promise<VectorizeVectorMutation> {
+    const res = await this._send(
+      Operation.VECTOR_UPSERT,
+      `indexes/${this.indexId}/upsert`,
+      {
+        method: "POST",
+        body: JSON.stringify({
+          vectors: vectors.map((vec) => ({
+            ...vec,
+            values: Array.isArray(vec.values)
+              ? vec.values
+              : Array.from(vec.values),
+          })),
+        }),
+        headers: {
+          "content-type": "application/json",
+          "cf-vector-search-dim-width": String(
+            vectors.length ? vectors[0]?.values?.length : 0
+          ),
+          "cf-vector-search-dim-height": String(vectors.length),
+          accept: "application/json",
+        },
+      }
+    );
+
+    return await toJson<VectorizeVectorMutation>(res);
+  }
+
+  public async getByIds(
+    options: VectorizeGetByIdsOptions
```

Comment:
I guess we could have this as separate `ids: number[], options: Options` params, and just post them to the API in a wrapper object

Reply:
(this would unify it with `query()` above)